### ISSUE TITLE
[HOLD #9581] [$500] IdentityForm

### ISSUE THREAD

**Issue by neil-marcellini** [Expensify/App#10729]
Coming from the [New Expensify Forms](https://docs.google.com/document/d/1DI4jQ0nyfi2JczhPT440orwlXsTP_O-aGvg-HLmcbwI/edit#heading=h.mnjguil21u0a) design doc, we should refactor the inputs within the [IdentityForm](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/IdentityForm.js) to be compatible with the new Form component, following the guidelines below:

Here's an example of a Form refactor: https://github.com/Expensify/App/pull/9056
[Here](https://github.com/Expensify/App/blob/8e07c62fb9780c627106f45eb72614466a850f27/src/components/AddressSearch.js#L190) is a component that had its inputs refactored to be compatible with Form.js.

## Guidelines
Contributors should first read [FORMS.md](https://github.com/Expensify/App/blob/b81ff69197e091727b33f8dd53fbc429fdd616b9/contributingGuides/FORMS.md).
The refactored inputs should be functional when wrapped by Form.js and when not.

1. Add an optional inputID prop. 
2. Add an optional shouldSaveDraft prop that defaults to false. 
3. Make the value prop optional.
4. Change any onChange prop, e.g. onTextChange to onChange.
5. In the input’s onBlur method, call props.onBlur().
6. In the input’s onChange method, call props.onChange() .
7. Remove the hasError prop.
8. Update the error message to display if errorText is truthy.
9. Make sure that props.ref is attached to the appropriate DOM node. This could involve forwarding the ref to a child component.
10. Make sure the ref is exposing both value and focus().
11. Add an optional maxLength prop to text inputs and UI to display the current character count out of the limit.
12. Add a hint prop and display the text under the input if there is no existing error.
13. Remove any unused code.

## Testing
Verify that:
- [ ] UI looks as it did before the refactor
- [ ] Values can be added and edited
- [ ] Errors are highlighted correctly (input border)
- [ ] Error messages show up correctly
- [ ] Draft values are saved properly
- [ ] Form alerts are displayed correctly
- [ ] Clicking the `fix the errors` link focuses the first input with error
- [ ] No duplicate submission of the form occurs (when it's already submitting)

**Comment by neil-marcellini** [comment]
It might be a good idea to price this issue at $500+ right away since it's a precursor to the RequestorStep refactor.

**Comment by mollfpr** [comment]
@neil-marcellini May I know the plan for deploying this form? Is this form will be deployed before refactoring `ACHContractStep` and `RequestStep`?

I also think the proposal should include refactoring `AddressForm` (since it's been used in `IdentityForm`).

**Comment by neil-marcellini** [comment]
Great point about the AddressForm. I created this issue to refactor it https://github.com/Expensify/App/issues/10738. I also updated this description because this refactor will only involve updating the inputs to be compatible with Form.js. Please submit your proposal for AddressForm. Thanks.

**Comment by neil-marcellini** [comment]
> deployed before refactoring `ACHContractStep` and `RequestStep`?

Yes. We will start with AddressForm, then IdentityForm, and then RequestorStep. Then eventually `ACHContractStep`.

**Comment by mollfpr** [comment]
@neil-marcellini Got it! I'm on my way to send the proposal for AddressForm.

**Comment by JmillsExpensify** [comment]
Just created an Upwork job here: https://www.upwork.com/jobs/~01604cc324bbf955b6. I went ahead and started the posting at $500. cc @neil-marcellini

**Comment by JmillsExpensify** [comment]
Proposals happening in https://github.com/Expensify/App/issues/10738. We're still on hold until that one goes through.

**Comment by luacmartins** [comment]
I'm ooo until Wednesday. Will come back to this when I'm back!

**Comment by luacmartins** [comment]
Still looking for proposals!

**Comment by Santhosh-Sellavel** [comment]
On hold @luacmartins

**Comment by Santhosh-Sellavel** [comment]
On hold

**Comment by Santhosh-Sellavel** [comment]
On hold

**Comment by JmillsExpensify** [comment]
Still on hold, though we're going to pick up the larger keyboard initiative soon.

**Comment by Santhosh-Sellavel** [comment]
Can we bump this to weekly while we wait?

**Comment by JmillsExpensify** [comment]
Still on hold per the larger keyboard initiative. And yes, will bump to weekly!

**Comment by luacmartins** [comment]
On hold

**Comment by JmillsExpensify** [comment]
Still on hold.

**Comment by luacmartins** [comment]
https://github.com/Expensify/App/issues/10738 is merged. Hold removed!

**Comment by mollfpr** [comment]
### Proposal

1. Updating the props to add `inputKeys`, `shouldSaveDraft`, and `translate` because we will need to remove `withLocalize` on `IdentityForm` to work with `Form.js`
2. Passing the `inputID` and `shouldSaveDraft` to the `TextInput` and `DatePicker` from `inputKeys` props
3. Passing `inputKeys` to `AddressForm`
4. Updating the components that use `IdentityForm` to pass `translate`

```diff
diff --git a/src/pages/ReimbursementAccount/IdentityForm.js b/src/pages/ReimbursementAccount/IdentityForm.js
index 7925133bb..5abfe005d 100644
--- a/src/pages/ReimbursementAccount/IdentityForm.js
+++ b/src/pages/ReimbursementAccount/IdentityForm.js
@@ -1,9 +1,9 @@
 import React from 'react';
 import {View} from 'react-native';
 import PropTypes from 'prop-types';
+import _ from 'underscore';
 import TextInput from '../../components/TextInput';
 import styles from '../../styles/styles';
-import withLocalize, {withLocalizePropTypes} from '../../components/withLocalize';
 import CONST from '../../CONST';
 import DatePicker from '../../components/DatePicker';
 import AddressForm from './AddressForm';
@@ -45,12 +45,39 @@ const propTypes = {
     /** Any errors that can arise from form validation */
     errors: PropTypes.objectOf(PropTypes.bool),

-    ...withLocalizePropTypes,
+    /** The map for inputID of the inputs */
+    inputKeys: PropTypes.shape({
+        firstName: PropTypes.string,
+        lastName: PropTypes.string,
+        street: PropTypes.string,
+        city: PropTypes.string,
+        state: PropTypes.string,
+        zipCode: PropTypes.string,
+        dob: PropTypes.string,
+        ssnLast4: PropTypes.string,
+    }),
+
+    /** Saves a draft of the input value when used in a form */
+    shouldSaveDraft: PropTypes.bool,
+
+    /** Returns translated string for given locale and phrase */
+    translate: PropTypes.func.isRequired,
 };

 const defaultProps = {
     style: {},
     values: {
+        firstName: undefined,
+        lastName: undefined,
+        street: undefined,
+        city: undefined,
+        state: undefined,
+        zipCode: undefined,
+        dob: undefined,
+        ssnLast4: undefined,
+    },
+    errors: {},
+    inputKeys: {
         firstName: '',
         lastName: '',
         street: '',
@@ -60,7 +87,7 @@ const defaultProps = {
         dob: '',
         ssnLast4: '',
     },
-    errors: {},
+    shouldSaveDraft: false,
 };

 const IdentityForm = (props) => {
@@ -73,6 +100,8 @@ const IdentityForm = (props) => {
             <View style={[styles.flexRow]}>
                 <View style={[styles.flex2, styles.mr2]}>
                     <TextInput
+                        inputID={props.inputKeys.firstName}
+                        shouldSaveDraft={props.shouldSaveDraft}
                         label={`${props.translate('common.firstName')}`}
                         value={props.values.firstName}
                         onChangeText={value => props.onFieldChange({firstName: value})}
@@ -81,6 +110,8 @@ const IdentityForm = (props) => {
                 </View>
                 <View style={[styles.flex2]}>
                     <TextInput
+                        inputID={props.inputKeys.lastName}
+                        shouldSaveDraft={props.shouldSaveDraft}
                         label={`${props.translate('common.lastName')}`}
                         value={props.values.lastName}
                         onChangeText={value => props.onFieldChange({lastName: value})}
@@ -89,6 +120,8 @@ const IdentityForm = (props) => {
                 </View>
             </View>
             <DatePicker
+                inputID={props.inputKeys.dob}
+                shouldSaveDraft={props.shouldSaveDraft}
                 label={`${props.translate('common.dob')}`}
                 containerStyles={[styles.mt4]}
                 placeholder={props.translate('common.dateFormat')}
@@ -98,6 +131,8 @@ const IdentityForm = (props) => {
                 maximumDate={new Date()}
             />
             <TextInput
+                inputID={props.inputKeys.ssnLast4}
+                shouldSaveDraft={props.shouldSaveDraft}
                 label={`${props.translate('common.ssnLast4')}`}
                 containerStyles={[styles.mt4]}
                 keyboardType={CONST.KEYBOARD_TYPE.NUMBER_PAD}
@@ -107,6 +142,7 @@ const IdentityForm = (props) => {
                 maxLength={CONST.BANK_ACCOUNT.MAX_LENGTH.SSN}
             />
             <AddressForm
+                inputKeys={_.omit(props.inputKeys, ['firstName', 'lastName', 'dob', 'ssnLast4'])}
                 translate={props.translate}
                 streetTranslationKey="common.personalAddress"
                 values={props.values}
@@ -120,4 +156,4 @@ const IdentityForm = (props) => {
 IdentityForm.propTypes = propTypes;
 IdentityForm.defaultProps = defaultProps;
 IdentityForm.displayName = 'IdentityForm';
-export default withLocalize(IdentityForm);
+export default IdentityForm;
```

```diff
diff --git a/src/pages/ReimbursementAccount/RequestorStep.js b/src/pages/ReimbursementAccount/RequestorStep.js
index 70e00a46b..67445d753 100644
--- a/src/pages/ReimbursementAccount/RequestorStep.js
+++ b/src/pages/ReimbursementAccount/RequestorStep.js
@@ -209,6 +209,7 @@ class RequestorStep extends React.Component {
                             </TextLink>
                         </View>
                         <IdentityForm
+                            translate={this.props.translate}
                             onFieldChange={this.clearErrorsAndSetValues}
                             values={{
                                 firstName: this.state.firstName,
```

```diff
diff --git a/src/pages/ReimbursementAccount/ACHContractStep.js b/src/pages/ReimbursementAccount/ACHContractStep.js
index a5b131922..f6e21955b 100644
--- a/src/pages/ReimbursementAccount/ACHContractStep.js
+++ b/src/pages/ReimbursementAccount/ACHContractStep.js
@@ -231,6 +231,7 @@ class ACHContractStep extends React.Component {
                                         {this.props.translate('beneficialOwnersStep.additionalOwner')}
                                     </Text>
                                     <IdentityForm
+                                        translate={this.props.translate}
                                         style={[styles.mb2]}
                                         onFieldChange={values => this.clearErrorAndSetBeneficialOwnerValues(index, values)}
                                         values={{
```

The usage `IdentityForm` to use with `Form.js` will be:
```
                        <IdentityForm
                            translate={this.props.translate}
                            shouldSaveDraft
                            inputKeys={{
                                firstName: 'firstName',
                                lastName: 'lastName',
                                city: 'requestorAddressCity',
                                state: 'requestorAddressState',
                                street: 'requestorAddressStreet',
                                zipCode: 'requestorAddressZipCode',
                                dob: 'dob',
                                ssnLast4: 'ssnLast4',
                            }}
                        />
```

**Comment by luacmartins** [comment]
@Santhosh-Sellavel let me know when you had the chance to review @mollfpr's proposal!

**Comment by JmillsExpensify** [comment]
@Santhosh-Sellavel I wanted to check in on this one. Do you think you'll be able to review @mollfpr's proposal today? Let's get this one kicked off now that we're unheld.

**Comment by JmillsExpensify** [comment]
Also for clarity, a bunch of time has passed since we originally created the Upwork job, so it's automatically closed in the meantime. I will create a fresh one just as soon as we have a proposal agreed upon.

**Comment by Santhosh-Sellavel** [comment]
@mollfpr what about this [focus-and-tab-behavior](https://github.com/Expensify/App/blob/b81ff69197e091727b33f8dd53fbc429fdd616b9/contributingGuides/FORMS.md#focus-and-tab-behavior)

**Comment by mollfpr** [comment]
@Santhosh-Sellavel Do you mean the option to change the order? Do we have an example of how we handle that case?

In this example, I'm modifying `RequestorStep` to work with `Form.js` and also using the proposed `IdentityForm`. In the video, the focus and tab behavior is working correctly in order.

https://user-images.githubusercontent.com/25520267/198495851-f5be4474-525c-4162-9700-5f4ebde3dc8f.mov

**Comment by Santhosh-Sellavel** [comment]
> @mollfpr what about this [focus-and-tab-behavior](https://github.com/Expensify/App/blob/b81ff69197e091727b33f8dd53fbc429fdd616b9/contributingGuides/FORMS.md#focus-and-tab-behavior)
> 
> 

But there isn't any implementation done for this, like described in the doc.

Will wait for @luacmartins to confirm.

**Comment by luacmartins** [comment]
@rushatgabhane I think what @mollfpr showed in their video is correct. From the docs `In most cases this can be achieved by composition, i.e. rendering the components in the correct order`, so no specific implementation is needed if the sequence is correct.

**Comment by Santhosh-Sellavel** [comment]
You tagged wrongly @luacmartins then we are good to go with @mollfpr proposal.

**Comment by luacmartins** [comment]
Oops sorry! Edited my comment

**Comment by mollfpr** [comment]
@luacmartins @Santhosh-Sellavel PR raised, thanks! 🚀

**Comment by JmillsExpensify** [comment]
@luacmartins @Santhosh-Sellavel mind prioritizing a review of the linked PR? Thank you!

**Comment by JmillsExpensify** [comment]
The linked PR is merged. @Santhosh-Sellavel @mollfpr I have created a fresh Upwork job since the prior one has expired. Please apply here: https://www.upwork.com/jobs/~01a9b108e6002bd2e7.

**Comment by Santhosh-Sellavel** [comment]
Done thanks @JmillsExpensify!

**Comment by JmillsExpensify** [comment]
Thank you! Hired.

**Comment by mollfpr** [comment]
@JmillsExpensify Applied too

**Comment by Santhosh-Sellavel** [comment]
Coming from this [issue](https://github.com/Expensify/App/issues/12400),
@mollfpr 
Our PR is Failing this test
> - [ ] Clicking the fix the errors link focuses the first input with error

Can you create a follow-up PR to address this one?

cc: @luacmartins

**Comment by Santhosh-Sellavel** [comment]
On clicking fix errors, first error input should be focused.

**Comment by mollfpr** [comment]
I do not understand what should I do...

But I just tested with one of the proposals on https://github.com/Expensify/App/issues/12400 is truly fixes the issue.

**Comment by JmillsExpensify** [comment]
@Santhosh-Sellavel @luacmartins Can one of you clarify where we're at? Are we waiting on a fix from @mollfpr in the linked PR?

**Comment by Santhosh-Sellavel** [comment]
@mollfpr I've just tested a proposal there it's not working on this form.

**Comment by JmillsExpensify** [comment]
Nice work getting this merged! I'll circle back in the coming days when the regression period has ended.

**Comment by Santhosh-Sellavel** [comment]
> @mollfpr I've just tested a proposal there it's not working on this form.
> 

@mollfpr bump

**Comment by mollfpr** [comment]
@Santhosh-Sellavel 

https://github.com/Expensify/App/issues/12400#issuecomment-1308050178 in this test you using `CompanyStep` where we haven't refactored this form to use `Form.js`.
https://github.com/Expensify/App/blob/9e3a5abef512f7bb7c56993aa3f8474d788f1e5b/src/pages/ReimbursementAccount/CompanyStep.js#L193
https://github.com/Expensify/App/blob/9e3a5abef512f7bb7c56993aa3f8474d788f1e5b/src/pages/ReimbursementAccount/ReimbursementAccountForm.js#L65-L67

While this https://github.com/Expensify/App/issues/12400 issue happens in `BankAccountManualStep` where we use `Form.js`. You should try to test it in any form that uses `Form.js`.
https://github.com/Expensify/App/blob/9e3a5abef512f7bb7c56993aa3f8474d788f1e5b/src/pages/ReimbursementAccount/BankAccountManualStep.js#L79-L85
https://github.com/Expensify/App/blob/9e3a5abef512f7bb7c56993aa3f8474d788f1e5b/src/components/Form.js#L230-L232

**Comment by Santhosh-Sellavel** [comment]
@mollfpr  How to test this `IdentityForm` against the same issue, Should we have to wait until `RequesterStep` is complete?

**Comment by mollfpr** [comment]
@Santhosh-Sellavel IMO we shouldn't. This refactoring doesn't impact the `inputRefs` state of `Form.js`.

The current implementation of `onFixTheErrorsLinkPressed` is to get the input refs from the order of key in `this.state.errors`.

This issue https://github.com/Expensify/App/issues/12400 occurred because on the validation function of `BankAccountManualStep` the first to validate is the `accountNumber` value. So that's why the focus input is the account number field. If we change the order of that validation the `routingNumber` is the first value to validate, the focus input will be correct to the routing number input.
https://github.com/Expensify/App/blob/9e3a5abef512f7bb7c56993aa3f8474d788f1e5b/src/pages/ReimbursementAccount/BankAccountManualStep.js#L36-L54

It's already explained in this [proposal](https://github.com/Expensify/App/issues/12400#issuecomment-1304789990).

**Comment by Santhosh-Sellavel** [comment]
@mollfpr  I believe you miss understood, I'm  not saying to put 12400 on hold until. 


I am asking, to test this  'IdentityForm' we should wait until 'RequestorStep' to test whether this works as expected or not against the issue as we didn't complete RequesterStep which uses Identity Form.

**Comment by mollfpr** [comment]
@Santhosh-Sellavel Ahh sorry!

**Comment by Santhosh-Sellavel** [comment]
So we have to wait right?

**Comment by mollfpr** [comment]
@Santhosh-Sellavel Let’s ask @neil-marcellini and @luacmartins

**Comment by neil-marcellini** [comment]
My understanding is that to test a solution to this issue https://github.com/Expensify/App/issues/12400, we need a form that uses IdentityForm. Is that indeed the problem in question? Could we use Storybook to test IdentityForm in isolation? If not then it sounds like we will have to hold https://github.com/Expensify/App/issues/12400 on https://github.com/Expensify/App/issues/9581.

**Comment by mollfpr** [comment]
I could create a storybook component for IdentityForm that use the `Form.js` but holding this until https://github.com/Expensify/App/issues/9581 passed is a good idea.

**Comment by JmillsExpensify** [comment]
@Santhosh-Sellavel @neil-marcellini The regression period has passed. At this point, this PR is merged and IndentityForm is done. What's our plan for testing this in conjunction with related issues? Should that happen in a separate issue when https://github.com/Expensify/App/issues/12400 and https://github.com/Expensify/App/issues/9581 are ready?

**Comment by Santhosh-Sellavel** [comment]
@JmillsExpensify When #9581 is ready, this should work!

**Comment by JmillsExpensify** [comment]
Ok, I'm going to go ahead and issue payment. Per the process, the PR is merged and has been for 7 days. I don't feel great about waiting on a related issue that is not yet merged. That part is on Expensify.

**Comment by JmillsExpensify** [comment]
Alright, all contributors paid out. I removed the `Awaiting Payment` label and I'm also updating the hold as well. We can come back to this and test when #9581 is ready.

**Comment by JmillsExpensify** [comment]
Review of the linked/holding PR is in progress! I'll have to check back in the coming days.

**Comment by JmillsExpensify** [comment]
#9581 is merged and on staging at this point, so I think we should be able to remove the merge and move forward with next steps. Does that sound right @Santhosh-Sellavel @neil-marcellini?

**Comment by Santhosh-Sellavel** [comment]
We can close this now or along with #9581.This is already merged, we were waiting for test a particular test case which is passing now. @JmillsExpensify

**Comment by JmillsExpensify** [comment]
Awesome, sounds like a plan. As for regression tests, I think we are already covered via [this test](https://expensify.testrail.io/index.php?/cases/view/1971185&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=229063). Accordingly, I'm going to close. cc @luacmartins please re-open if you disagree and I'll make sure any additional cases are added to that existing test.
